src/order/closure.lean

```diff
@@ -0,0 +1,75 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Bhavik Mehta
+-/
+import order.basic
+import order.preorder_hom
+import tactic.monotonicity
+
+/-!
+# Closure operators on a partial order
+
+We define (bundled) closure operators on a partial order as an monotone (increasing), extensive
+(inflationary) and idempotent function.
+We define closed elements for the operator as elements which are fixed by it.
+-/
+universe u
+
+variables (α : Type u) [partial_order α]
+
+/--
+A closure operator on the partial order `α` is a monotone function which is extensive (every x
+is less than its closure) and idempotent.
+-/
+@[ext]
+structure closure_operator extends α →ₘ α :=
+(extensive' : ∀ x, x ≤ to_fun x)
+(idem' : ∀ x, to_fun (to_fun x) = to_fun x)
+
+instance : has_coe_to_fun (closure_operator α) :=
+{ F := _, coe := λ c, c.to_fun }
+
+initialize_simps_projections closure_operator (to_fun → apply)
+
+namespace closure_operator
+
+/-- The identity function as a closure operator. -/
+@[simps]
+def id : closure_operator α :=
+{ to_fun := λ x, x,
+  monotone' := λ _ _ h, h,
+  extensive' := λ _, le_refl _,
+  idem' := λ _, rfl }
+
+instance : inhabited (closure_operator α) := ⟨id α⟩
+
+variables {α} (c : closure_operator α)
+
+@[mono] lemma monotone : monotone c := c.monotone'
+lemma extensive (x : α) : x ≤ c x := c.extensive' x
```

Comment:
I'm fine with naming the field of the definition `extensive` or `extensive'`, but I'm not sure it's a recognizable enough term for it to be the only name of this lemma.

Reply:
Good point - is `le_closure` better?

Reply:
I think that's guessable.